File: Top 100 Codes/06-Advanced Codes related to Arrays/08-Find duplicate in an array of N+1 Integers/findduplicateinanarrayofN+1Integers.py
def find_duplicate(nums):
    seen = set()
    for num in nums:
        if num in seen:
            return num
        seen.add(num)
    return -1  # If no duplicate found (shouldn't happen for N+1 array)

nums = [1, 3, 4, 2, 2]
print("Duplicate number is:", find_duplicate(nums))

File: Top 100 Codes/06-Advanced Codes related to Arrays/08-Find duplicate in an array of N+1 Integers/findduplicateinanarrayofN+1Integers.cpp
#include <iostream>
#include <vector>
using namespace std;

int findDuplicate(const vector<int>& nums) {
    int slow = nums[0];
    int fast = nums[0];

    // Phase 1: Detect cycle
    do {
        slow = nums[slow];
        fast = nums[nums[fast]];
    } while (slow != fast);

    // Phase 2: Find entrance to the cycle (duplicate)
    fast = nums[0];
    while (slow != fast) {
        slow = nums[slow];
        fast = nums[fast];
    }

    return slow;
}

int main() {
    vector<int> nums = {1, 3, 4, 2, 2};  // Example input with duplicate 2
    int duplicate = findDuplicate(nums);
    cout << "Duplicate number is: " << duplicate << endl;
    return 0;
}
